pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1128 @@
+	/// * a) If we increase by 500, we can reduce the `foreign_swap.amount`
+	///   fully, which we denote by adding the 500 to the `foreign_done` amount.
+	///   Moreover, we can immediately invest the 500. The resulting state is
+	///   `(done_amount = 1500, investing = 500)`.
+	/// * b) If we increase by 400, we can reduce the `foreign_swap.amount` only
+	///   by 400 and increase both the `investing` as well as `foreign_done`
+	///   amount by that. The resulting state is
+	/// `(done_amount = 1400, foreign_swap.amount = 100, investing = 400)`.
+	/// * c) If we increase by 600, we can reduce the `foreign_swap.amount`
+	///   fully and need to add a swap into pool currency for 100. Moreover both
+	///   the `investing` as well as `foreign_done` amount can only be increased
+	///   by 500. The resulting state is
+	/// `(done_amount = 1500, pool_swap.amount = 100, investing = 500)`.
+	///
+	/// NOTES:
+	/// * We can never directly compare `swap.amount` and `invest_amount` with
+	///   `foreign_swap.amount` and `done_amount` if the currencies mismatch as
+	///   the former pair is denominated in pool currency and the latter one in
+	///   foreign currency.
+	/// * We can ignore handling all states which include `*SwapIntoForeignDone`
+	///   without `ActiveSwapIntoForeignCurrency*` as we consume the done amount
+	///   and transition in the post transition phase. To be safe and to not
+	///   make any unhandled assumptions, we throw `DispatchError::Other` for
+	///   these states though we need to make sure this can never occur!
+	fn handle_increase(
+		&self,
+		swap: Swap<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_increase_non_foreign(self, swap);
+		}
+
+		match &self {
+			Self::NoState => Ok(Self::ActiveSwapIntoPoolCurrency { swap }),
+			// Add pool swap
+			Self::InvestmentOngoing { invest_amount } => {
+				Ok(Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+					swap,
+					invest_amount: *invest_amount,
+				})
+			}
+			// Bump pool swap
+			Self::ActiveSwapIntoPoolCurrency { swap: pool_swap } => {
+				swap.ensure_currencies_match(pool_swap, true)?;
+				Ok(Self::ActiveSwapIntoPoolCurrency {
+					swap: Swap {
+						amount: swap.amount.ensure_add(pool_swap.amount)?,
+						..swap
+					},
+				})
+			}
+			// Reduce foreign swap amount by the increasing amount and increase investing amount as
+			// well adding foreign_done amount by the minimum of active swap amounts
+			Self::ActiveSwapIntoForeignCurrency { swap: foreign_swap } => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let foreign_amount_pool_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_in,
+					swap.currency_out,
+					foreign_swap.amount,
+				)?;
+				let pool_amount_foreign_denominated = T::CurrencyConverter::stable_to_stable(
+					swap.currency_out,
+					swap.currency_in,
+					swap.amount,
+				)?;
```

Comment:
We would have to adjust at least the `SimpleStableCurrencyConverter` impl of `SimpleCurrencyConversion` then and maybe already now the `SimpleCurrencyConversion`. I did not want to waste too much energy into this before receiving a review until finishing integration tests because there is no optimal solution at hand and for the time being, we will be fine with having one-to-one conversion. 

Right now, `SimpleCurrencyConversion` just exposes `stable_to_stable`. There was an earlier variant of this which exposed `foreign_to_pool` and the opposite which is more appropriate. The `SimpleStableCurrencyConverter` impl of `SimpleCurrencyConversion` implicitly assumes the provided conversion ratio to be bidirectional, i.e. to be one. The reason is that else we need to hardcode conversion ratios for specific currency pairs. Of course, that's much better than implicitly assuming bidirectional `identity` conversions (apart from the re-denomination) but also overhead if we don't plan to tackle this issue in the near future.

WDYT?

Reply:
I am fine with the intermediate solution. Was just curious. I would not adapt this in this version.